R/select_nse.R

```diff
@@ -0,0 +1,458 @@
+  columns <- colnames(data)
+
+  # avoid conflicts
+  conflicting_packages <- .conflicting_packages("poorman")
+  on.exit(.attach_packages(conflicting_packages))
+
+  expr_select <- substitute(select, env = parent.frame(1L))
+  expr_exclude <- substitute(exclude, env = parent.frame(1L))
+
+  # when exclude is not an argument called from the function (e.g data_to_long),
+  # do not consider "exclude" as a symbol
+  if (deparse(expr_exclude) == "exclude" && is.null(substitute(exclude))) {
+    expr_exclude <- NULL
+  }
+
+  # directly return all names if select == exclude == NULL
+  if (is.null(expr_select) && is.null(expr_exclude)) {
+    return(columns)
+  }
+
+  # get the position of columns that are selected or excluded
+  selected <- .eval_expr(expr_select, data,
+    ignore_case = ignore_case,
+    regex = regex, verbose
+  )
+  excluded <- .eval_expr(expr_exclude, data,
+    ignore_case = ignore_case,
+    regex = regex, verbose
+  )
+
+  if ((any(selected < 0) && any(selected > 0)) ||
+    (any(excluded < 0) && any(excluded > 0))) {
+    insight::format_error(
+      "You can't mix negative and positive indices in `select` or `exclude`."
+    )
+  }
+
+  # variable positions -> variable names
+  selected <- columns[selected]
+  excluded <- columns[excluded]
+
+  if (length(selected) == 0L) {
+    if (length(excluded) == 0L) {
+      out <- character(0L)
+    } else {
+      out <- setdiff(columns, excluded)
+    }
+  } else {
+    out <- setdiff(selected, excluded)
+  }
+
+  out
+}
+
+
+# This is where we dispatch the expression to several helper functions.
+# This function is called multiple times for expressions that are composed
+# of several symbols/language.
+#
+# Ex:
+# * "cyl" -> will go to .select_char() and will return directly
+# * cyl:gear -> function (`:`) so find which function it is, then get the
+#   position for each variable, then evaluate the function with the positions
+
+.eval_expr <- function(x, data, ignore_case, regex, verbose) {
+  if (is.null(x)) {
+    return(NULL)
+  }
+
+  type <- typeof(x)
+
+  out <- switch(type,
+    "integer" = x,
+    "double" = as.integer(x),
+    "character" = .select_char(data, x, ignore_case, regex = regex, verbose),
+    "symbol" = .select_symbol(data, x, ignore_case, regex = regex, verbose),
+    "language" = .eval_call(data, x, ignore_case, regex = regex, verbose),
+    insight::format_error(
+      "Expressions of type <", typeof(x),
+      "> cannot be evaluated for use when subsetting."
+    )
+  )
+
+  out
+}
+
+
+# Possibilities:
+# - quoted variable name
+# - quoted variable name with ignore case
+# - character that should be regex-ed on variable names
+# - special word "all" to return all vars
+
+.select_char <- function(data, x, ignore_case, regex, verbose) {
+  # use colnames because names() doesn't work for matrices
+  columns <- colnames(data)
+  if (isTRUE(regex)) {
+    grep(x, columns)
+  } else if (length(x) == 1L && x == "all") {
+    return(seq_along(data))
```

Comment:
Why do we only return from here and from no other branches of the conditional expression?